# CompAtlas: segment teleost genomes by post-TGD ancestral chromosomes

 [![Snakemake](https://img.shields.io/badge/snakemake-≥5.5.4-brightgreen.svg)](https://snakemake.bitbucket.io) ![Snakemake-Report](https://img.shields.io/badge/snakemake-report-green.svg)

## Table of content

  - [Description](#description)
  - [Usage](#usage)
  - [References](#references)

## Description

CompAtlas builds maps of TGD-duplicated genomes, annotating them with respect to ancestral chromosomes generated by the TGD.

The pipeline takes as 4 inputs: (1) predefined pre-TGD ancestral chromosomes mapped on a subset of 4 input genomes ("references"), (2) genes coordinates files for all studied teleosts, (3) gene trees with all studied teleost, (4) the corresponding species tree. CompAtlas assigns post-duplication chromosomes to genes of all TGD-duplicated genomes present in the gene trees.

The major steps are the following:

- Define TGD-paralogous segments independently for each of the 4 genomes used as reference, to obtain post-duplication chromosomes annotations for reference genomes,
- Homogenize the nomenclature across the 4 reference teleosts, so that orthologous regions have the same the post-duplication chromosomes name (i.e. 'a' and 'b' may be swapped for consistency),
- Use a majority vote of orthologous genes in the references to solve conflicts,
- Propagate the annotatioins to genes of all duplicated genomes using orthologous relations in gene trees


## Usage

All dependencies are listed in `envs/comp_atlas.yaml` (mainly snakemake, ete3, matpotlib and seaborn).

- Create the conda environment (this may take a few minutes):
```
conda install mamba
mamba env create -f envs/comp_atlas.yaml
```

- Activate the conda environment:
```
conda activate comparative_atlas
```

- Run on toy example data:
```
snakemake --configfile config_map_ens89.yaml --cores 4
```

- Generate a snakemake report after a run:

- To run on a user-defined dataset, create a new config file following the provided example. Provided with the pre-duplication ancestral karyotype, compAtlas can run on any set of WGD-descended genomes.

## References

CompAtlas takes as input the pre-defined pre-TGD ancestral chromosomes from:

- [(Nakatani and McLysaght 2017)](https://academic.oup.com/bioinformatics/article/33/14/i369/3953974): Nakatani Y, McLysaght A. 2017. Genomes as documents of evolutionary history: a probabilistic macrosynteny model for the reconstruction of ancestral genomes. Bioinformatics 33:i369–i378.
